File: nodejs/business-context/src/expression/FunctionCall.js
export default class FunctionCall {
    constructor(methodName, type, args) {
        this.methodName = methodName
        this.type = type
        this.args = args
    }

    interpret(context) {
        const interpretedArgs = this.args.map(arg => arg.interpret(context))
        let func = context.getFunction(this.methodName)
        return func.apply(func, interpretedArgs)
    }

    getName() {
        return 'function-call'
    }

    getArguments() {
        return this.args
    }

    getProperties() {
        return {
            methodName: this.methodName,
            type: this.type.name,
        }
    }

    toString() {
        return 'call ' + this.methodName + '()'
    }
}

File: java/business-context/src/main/java/cz/filipklimes/diploma/framework/businessContext/expression/FunctionCall.java
package cz.filipklimes.diploma.framework.businessContext.expression;

import cz.filipklimes.diploma.framework.businessContext.weaver.BusinessOperationContext;
import cz.filipklimes.diploma.framework.businessContext.Function;

import java.io.*;
import java.util.*;

public class FunctionCall<T> implements Expression<T>, Serializable
{

    private final String methodName;
    private final ExpressionType type;
    private final Expression<?>[] arguments;

    public FunctionCall(final String methodName, final ExpressionType type, final Expression<?>... arguments)
    {
        this.methodName = methodName;
        this.type = type;
        this.arguments = arguments.clone();
    }

    @Override
    @SuppressWarnings("unchecked")
    public T interpret(final BusinessOperationContext context)
    {
        Function function = context.getFunction(methodName);

        Object[] calculatedArguments = Arrays.stream(arguments)
            .map(argument -> argument.interpret(context))
            .toArray();

        return (T) type.getUnderlyingClass().cast(function.execute(calculatedArguments));
    }

    @Override
    public Collection<Expression<?>> getArguments()
    {
        return Arrays.asList(arguments);
    }

    @Override
    public Map<String, String> getProperties()
    {
        Map<String, String> propertyMap = new HashMap<>();
        propertyMap.put("methodName", methodName);
        propertyMap.put("type", type.getName());
        return propertyMap;
    }

    @Override
    public String getName()
    {
        return "function-call";
    }

    @Override
    public void accept(final ExpressionVisitor visitor)
    {
        visitor.visit(this);
    }

    @Override
    public String toString()
    {
        return String.format("call %s()", methodName);
    }

}
